APA overhaul

bwiernik opened the pull request:
- Added support for interviews, archival material, and audiovisual materials
- Updated the code for the (description) and [format] blocks to accommodate the various item cases and ensure that only one of each gets printed for any given reference

I've done testing, but I need to do more checking still. Is there a good set of test items available to start with to check that all of the conditionals are working? I can make my own test suite, but if there is a good starting set that will save a lot of time.

bwiernik commented:
Happy to discuss some of the decisions that had to be made. Specific areas that were challenging were accommodating:

1. Archival and non-archival manuscripts
1. Personal communications that are published/available versus not
1. Reports and songs that have container-titles versus don't

I added a few additional hard-coded strings (marked as in need of localization). Most of them should come up fairly rarely I suspect, but we can consider some alternative options if we want to break localization less until more terms can be added to CSL.

bwiernik commented:
Question:

Is there any way to get a `names` element to render multiple creator types together, with the order presented in the item data, rather than separated by creator type? For example, if a book has two authors and an illustrator, then this:
`<names variable="author illustrator" delimiter=", ">`

Produces:
Author, F., & Author, S., Illustrator, F. (Ill.)

Rather than:
Author, F., Author, S., & Illustrator, F. (Ill.)

adam3smith commented:
No -- every `name` call (i.e. every creator type) is being handled separately. I have to say that I find the second example really confusing, too, with no way of telling who the (Ill) refers to.

bwiernik commented:
Here’s a more common example:

If I have a chapter in a book with an editor and a translator (different people), then the existing APA style renders as:
In A. S. Smith (Ed.), T. Jones (Trans.), _Book_

Rather than with the ampersand:
In A. S. Smith (Ed.) & T. Jones (Trans.), _Book_

This CSL behavior seems very counter intuitive to me. Particularly frustrating is that `and` can only be applied to `name` and not to `names`, so it is impossible to create continuous lists with multiple creator types. For example, it’s impossible to correctly format editor and translator combinations or to even approach a correct rendering for media creators:
Allen, A. (Director), Berry, B. (Director), & Collins, C. (Producer)

bwiernik commented:
(But thanks for the clarification!)

bwiernik commented:
@adam3smith Okay, I've run through a bunch of tests, and everything is looking good on my end. Once you've had a chance to review, I will port the changes over to the other APA versions.

adam3smith commented:
Oh, and of course thanks for doing this in the first place. (And so far I've only reviewed the APA changes.

bwiernik commented on apa.csl:
```diff
@@ -249,72 +447,199 @@
-            <text variable="version"/>
+            <text variable="genre" text-case="title"/>
+            <!-- This should be localized -->
+            <number variable="number" prefix="No. "/>
```
Thanks. Lots of report numbers won’t be numeric, I think. Things like TR2346-A, so testing for numeric won’t work (and the manual has examples indicating that No. should used in those cases).

bwiernik replied:
Should a separate `number` term be created in a future CSL version?

bwiernik commented on apa.csl:
```diff
@@ -50,94 +51,296 @@
-            <label form="short" prefix=" (" text-case="title" suffix=")"/>
-          </names>
-        </group>
+      <if type="broadcast chapter entry entry-dictionary entry-encyclopedia graphic map paper-conference personal_communication report" variable="container-title" match="all">
```
citeproc treats this as testing for “type is one in this list and has a container title”. There are several styles in the repo with multiple types and match=“all”. Is that not the way it should be implemented in citeproc?

adam3smith replied:
yeah, that's definitely not correct. 
> “all” - (default), element only tests “true” when all conditions test “true” for all given test values

is unambiguous in that respect.
We tried to eliminate these as much as possible in 
https://github.com/citation-style-language/styles/pull/3407
CC @fbennett

bwiernik commented on apa.csl:
```diff
@@ -155,88 +358,83 @@
+          <text variable="URL"/>
+        </group>
+      </else-if>
+      <else-if variable="archive" type="article dataset report thesis" match="all">
```
This has been how citeproc-js has worked for years, and several other styles use type=“aaa bbb” match=“all”. citeproc-js seems to treat multiples in type=“ “ as meaning “type is one in this list”.

adam3smith replied:
I'm pretty sure that pandoc-citeproc does this correctly, though, so just accepting this isn't really an option. If we have remaining styles with this, we'll need to fix them. That said:

> and several other styles use type=“aaa bbb” match=“all”

Do you have examples? I can't find any.

bwiernik replied:
(Just to be clear, I'm not trying to argue with you, just being clear about why I wrote it that way [because it worked in citeproc-js, there were a few examples, and it seemed possible that type="" was intended to be evaluated as "type is one of these"]. Obviously it's not a big deal to add the four lines of code if this behavior is unintended.)

The styles that still have multiple types and match="all" are:
```
harvard-university-for-the-creative-arts.csl
  221,16:       <else-if type="report entry-dictionary entry-encyclopedia motion_picture chapter speech song paper-conference article-journal book" match="all" variable="URL">

lluelles-no-ibid.csl
  732,30:           <if locator="page" type="chapter article-magazine article-newspaper" match="all">
  742,20:           <else-if type="legislation bill" locator="section" match="all">

lluelles.csl
  757,30:           <if locator="page" type="chapter article-magazine article-newspaper" match="all">
  767,20:           <else-if type="legislation bill" locator="section" match="all">
  854,46:                   <if variable="title-short" type="legislation bill entry-dictionary" match="all">

mcgill-fr.csl
  918,20:           <else-if type="legislation bill" locator="section" match="all">
  977,46:                   <if variable="title-short" type="legislation bill entry-dictionary" match="all">

universite-de-montreal-faculte-de-musique.csl
  621,20:           <else-if type="legislation bill" locator="section" match="all">
  680,46:                   <if variable="title-short" type="legislation bill entry-dictionary" match="all">

associacao-brasileira-de-normas-tecnicas-unirio-eipp.csl
  61,11:       <if match="all" type="book chapter">
  86,11:       <if match="all" variable="edition" type="book chapter">
  145,15:           <if match="all" variable="volume" type="book chapter article-journal thesis">
  161,15:           <if match="all" variable="page" type="chapter article-journal">
  169,15:           <if match="all" variable="number-of-pages" type="book thesis">

journal-of-the-warburg-and-courtauld-institutes.csl
  480,16:       <else-if match="all" type="webpage post-weblog">
```

bwiernik commented on apa.csl:
```diff
@@ -50,94 +51,296 @@
+                <text variable="title" form="short" font-style="italic"/>
+              </else>
+            </choose>
+            <text macro="format"/>
```
APA style says that for items that don’t have a title, to provide a description in square brackets. For example, [Unpublished survey data] or [Photograph of a man in a hat]. This structure accommodates the user entering the brackets directly into title or omitting title entirely and instead using genre or medium to store this information.

bwiernik commented on apa.csl:
```diff
@@ -50,94 +51,296 @@
+        <choose>
+          <if variable="container-author">
+            <group>
+              <names variable="container-author" delimiter=", &amp; ">
```
Thanks, that was a clean-up oversight.

bwiernik commented on apa.csl:
```diff
@@ -50,94 +51,296 @@
+                <names variable="translator">
+                  <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
+                    <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
+                    <substitute>
```
The thought here was this: If there is a container-title and no author, then it's reasonable to assume that the "translator" refers to the translator of the chapter, rather than the container as a whole (i.e., the translator of a chapter in an edited book). So what this `choose` is doing is changing the relative order of translator and editor depending on if there is a container title or not. I can see a way to streamline this though. 

Two questions:
1. Is there a way to test for `editortranslator` (the editor and translator being the same)? The goal being to show the `editortranslator` label if applicable, but then to show only one of `editor` or `translator` if it's not.

1. Currently, if the `label` formatting is changed during a `names` element (e.g., in the current code, it starts with no label for author so that illustrator and composer will be shown without labels, then does long labels for director, then switches to short labels for editor/translator), the label formatting always conforms to the last label format it encounters. So, for example, if I have:
```<names variable="author" delimiter=", ">
          <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
          <substitute>
            <names variable="illustrator"/>
            <names variable="composer"/>
            <names variable="director">
              <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
              <label form="long" prefix=" (" suffix=")" text-case="capitalize-first"/>
            </names>
            <names variable="editor">
              <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
              <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
            </names>
            <names variable="editorial-director"/>
            <names variable="translator"/>
          </substitute>
</names>
```
then editorial-director and translator will be rendered by citeproc-js with short labels (rather than no labels). This seems like it isn't the correct behavior, correct? To get short labels with editorial-director and translator, I should need to use a complete names-name-label construct for each, correct?

bwiernik replied:
Also, do you have a preference for either of these?

1. No label given for `author`, but then a label specified for each of `director`, `translator` (conditional), `editor`, `editorial-director`, and `translator` (6 name lines):
```
<names variable="author" delimiter=", ">
  <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
  <substitute>
    <names variable="illustrator"/>
    <names variable="composer"/>
    <names variable="director">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
      <label form="long" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <choose>
      <if variable="container-title">
        <names variable="translator">
          <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
            <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
        </names>
      </if>
    </choose>
    <names variable="editor">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
        <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <names variable="editorial-director">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
        <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <names variable="translator">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
        <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <choose>
      <if type="report">
        <text variable="publisher"/>
      </if>
    </choose>
    <text macro="title"/>
    <text macro="format"/>
  </substitute>
</names>
```

2. Short label given for `author`, but then a different label specified for `illustrator` (empty), `composer` (empty), and `director` (long) (4 name lines):
```
<names variable="author" delimiter=", ">
  <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
  <label form="short" prefix=" (" suffix=")" text-case="capitalize-first"/>
  <substitute>
    <names variable="illustrator">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
    </names>
    <names variable="composer">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
    </names>
    <names variable="director">
      <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
      <label form="long" prefix=" (" suffix=")" text-case="capitalize-first"/>
    </names>
    <choose>
      <if variable="container-title">
        <names variable="translator"/>
      </if>
    </choose>
    <names variable="editor"/>
    <names variable="editorial-director"/>
    <names variable="translator"/>
    <choose>
      <if type="report">
        <text variable="publisher"/>
      </if>
    </choose>
    <text macro="title"/>
    <text macro="format"/>
  </substitute>
</names>
```

The second is somewhat shorter, but also less readable.

bwiernik commented:
@adam3smith I made most of the necessary changes. I have a few comments/questions on your notes above when you have a chance.

adam3smith commented:
how about my question re: testing library. Do you have one?

bwiernik commented:
Send you a Zotero group invite. It's a mix of items I made with various combinations of variables, as well as Emiliano's BBT test library.

bwiernik commented:
@adam3smith APA style wants container titles for treaties and other legal citations to be abbreviated. But specifying `container-title form="short"` trigger's Zotero's automatic MEDLINE journal abbreviation message in the Document Preferences window. That's obviously way too confusing for users. I would expect any database or user to probably supply the legal container titles in abbreviated form already (is that correct?). Given that and the rarity of these types of citations, I'm inclined to just enter these as `container-title` and expect that users will supply the short form. Does that sound reasonable?

bwiernik commented:
Update: I've added all of the examples from the APA manual and guide to electronic references to my test library. I've fixed all of the bugs I ran into as I was cleaning those up. I will run through a final check of the 250 items to be sure they all still render correctly.

bwiernik commented:
Okay, all 251 items in the test library render correctly, or as correctly as they reasonably can. https://www.zotero.org/groups/2205533/test_items_library

In the library, items that produce output that deviates from the APA manual are tagged with `incorrect output`. Of those, `CSL change` indicates changes to CSL that are needed to meet the formatting requirements. `weird APA` notes places where APA is inconsistent examples or rules that are too difficult to be able to implement effectively.

bwiernik commented:
@adam3smith Would it be helpful to schedule a call to talk through this style?

bwiernik commented:
@agruber reached out and is also interested in helping out here.

adam3smith commented:
This is next up on my list to review. I figured what I'd do is go through this, including with the test library, and flag anything that I think is unreasonable or too fragile (as well, of course, as anything that I think is actually wrong). If there's a lot, jumping on a phone call will probably be most efficient, if there's just a couple of things we can hash them out here for transparency. 
Thanks for doing this!

bwiernik commented on apa.csl:
```diff
@@ -50,94 +51,296 @@
+            </group>
+          </if>
+          <else>
+            <group>
```
FYI, on further consideration, I switched this to put the `container-author` in parentheses and retain the editor/translator in the `In` position.

adam3smith commented:
OK, I'm about 1/3 through, looking very good so far, I only have a handful of notes. Will send all together once I'm done.

bwiernik commented on apa.csl:
```diff
@@ -41,103 +41,357 @@
+            </choose>
+            <names variable="editor translator" delimiter=", ">
+              <name name-as-sort-order="all" and="symbol" sort-separator=", " initialize-with=". " delimiter=", " delimiter-precedes-last="always"/>
+                <label form="short" prefix=" (" suffix=")" text-case="title"/>
```
Illustrator and composer generally don't get labels in APA style. Editor and translator labels get short form. Other creator types get long form.

bwiernik commented on apa.csl:
```diff
@@ -146,243 +400,466 @@
-        <text variable="genre"/>
-        <text term="issue" form="short" text-case="capitalize-first"/>
-      </group>
+      <text variable="genre"/>
```
I didn't change this at all. APA style wants `U.S. patent No. 12345`.

I agree that `authority` would be good for the country—importing from Google Patents puts the full "United States" in `issuingAuthority`, PATO.gov leaves it empty; we can rely on users to shorten to "US" if desired.

That leaves out the word "patent". We could add change it year to show both `authority` then either `genre` (if present) or hard-coded text "patent" (and replace with `term="patent"` when available), like we do with `dataset`.

bwiernik replied:
The reason to include `genre` here would be to accommodate things like trademark applications, which have the same citation form.

adam3smith replied:
How about something like this (conceptually):
```
<text variable="authority"/>
<group delimiter=" ">
  <choose>
    <if variable="genre">
      <text variable="genre"/>
    </if>
    <else>
      <text value="patent"/>
    </else>
  </choose>
  <group delimter=" ">
    <text term="issue" form="short"/>
    </text variable="number"/>
  </group>
</group>
```
This follows the same logic we do for presentations and personal communication. And then handle the bibliography in a similar way.

bwiernik commented on apa.csl:
```diff
@@ -146,243 +400,466 @@
-            <text variable="DOI" prefix="https://doi.org/"/>
-          </if>
-          <else>
+          <if type="article article-journal article-magazine article-newspaper dataset paper-conference report speech thesis" match="any">
```
This is actually the structure for retrieving items stored in electronic databases, like Proquest Dissertations or ERIC. The test for `archive-place` is to distinguish that use of `archive`/`archive_location` from using those fields to refer to physical items in historical archives.

Rendering of physical archive information happens using the `archive` macro called in the `publisher` macro.

bwiernik commented on apa.csl:
```diff
@@ -146,243 +400,466 @@
+                <text variable="publisher"/>
+              </if>
+              <else>
+                <text variable="authority"/>
```
The example reference in the manual is:
Smith, I. M. (1988). _US.S. Patent No. 123,445._ Washington, DC: U.S. Patent and Trademark Office.

The "Washington, DC: U.S. Patent and Trademark Office" bit is intended to refer to "the official source from which the patent information can be retrieved". I actually think that `authority` is inappropriate to include there (i.e., there is no reason that "U.S. Patent and Trademark Office" should be stored in `authority`).

bwiernik replied:
@adam3smith You okay with dropping authority here entirely then?

adam3smith replied:
Wouldn't it be needed for the U.S. in U.S. Patent just as in the citation? I'm definitely OK to not use it for Washington, DC: U.S. Patent and Trademark Office. (though looking back me doing it this way may have been with Mendeley in mind?)

bwiernik replied:
Sorry, I meant to remove it from the publisher macro because authority is being used in the patent number macro above. Mendeley doesn’t use authority at all, maybe it was another client? 

One other question—patents stand out as the only type in the APA Manual that don’t use a DOI/URL in place of the publisher name/location if available. Patents aren’t mentioned in the 2012 update focused on electronic references. 

Should (1) the URL or DOI be included if no publisher is present, and (2) a URL or DOI take precedence over publisher?

bwiernik replied:
Ah, it looks like you wrote it this way with Zotero in mind—Zotero doesn’t have either publisher or publisher place on patent.

bwiernik replied:
Okay, APA Style has indicated that it is acceptable to replace the publisher/place with URL or DOI for patents consistent with other item types.

adam3smith commented:
I think we have everything covered then, right? Ready to pull the trigger?

bwiernik commented:
I will copy the changes to the other APA styles this afternoon. Then, we can pull the trigger.

bwiernik commented:
Okay, @adam3smith, let's pull the trigger!
